Introduce a warning timeout for MQTT topics

gbeine opened the pull request:
The feature allows to define a parameter 'timeout' for each topic. The timeout is specified in seconds.

Instead of notifying targets when a message arrives for a topic, mqttwarn will trigger the targets if no message is received on the topic for the specified period of time.

That's the first version of this feature, so any feedback is welcome.

Commit:
Introduce a warning timeout for MQTT topics

The feature allows to define a parameter 'timeout' for each topic.
The timeout is specified in seconds.

Instead of notifying targets when a message arrives for a topic, mqttwarn will trigger the targets if no message is received on the topic for the specified period of time.

Commit:
Fixed log levels and messages

Commit:
Minor improvements

Reduced invasiveness
- removed message_to_targets_handler by passing send_to_targets as on_timeout callback
- introduced notify_only_on_timeout switch for allowing notifications for arriving messages, too

Commit:
Add support for timeouts on wildcard topics

amotl commented:
Hi Gerrit,

apologies for the late reply. That's a nice liveness sensor you are contributing here. I didn't have a look at the details yet, but if @jpmens or @sumnerboy12 don't have any objections about adding that as a new kind of feature, let's try to bring it into mainline.

With kind regards,
Andreas.

gbeine commented:
@amotl Thank you - and no problem, we're all busy folks ;-)

I'd welcome any feedback as I'm not sure if I've done the implementation in the most effective way.
All I can say is, that it runs on some Raspberry Pi 2 and 4 machines with about 70 to 150 topics stable for months.

Commit:
Fix notify_only_on_timeout behavior

If notify_only_on_timeout is set, mqttwarn will send a notification in case the timeout was reached.
With this change, it will also send a message in case a message has been received within the timeout for the first time.
The assumption is, that only timeout/no timeout is relevant and if messages are received within the timeout, that's not a status change.
If all messages are of relevant, just set notify_only_on_timeout = False which is the default.

Commit:
Optimized last_state_timeout behavior

Make last_state_timeout private as no access from outside in required.
Initialize last_state_timeout with True as default to allow proper handling of the first messaged received even if no timeout was reached before.

gbeine commented:
Is there a chance to merge this PR?

jpmens commented:
I think @amotl wanted to comment on some of the design and implementation, but he's not yet had an opportunity to do so.

gbeine commented:
Thanks for the fast answer @jpmens!
It would be great to see how the feature can be improved.
Meanwhile I have to maintain an own docker image until the feature is merged ;-)

amotl commented:
Hi again. Thanks for the reminder. I will try to come back to this soon. 🙏

amotl commented:
> While your implementation looks pretty solid, I may have a few suggestions about design and implementation details as we go.

I think we will be happy to merge as-is, if nobody else has any objections. Maybe add software tests and a changelog item, when possible?
